torch/lib/c10d/reducer.cpp

```diff
@@ -429,6 +431,69 @@ void Reducer::set_divide_factor() {
   }
 }
 
+void Reducer::delay_all_reduce() {
+  std::lock_guard<std::mutex> lock(this->mutex_);
+
+  if (!static_graph_ || num_iterations_ > 1) {
+    return;
+  }
+
+  if (should_collect_runtime_stats()) {
+    record_backward_compute_end_time();
+  }
+  // The autograd engine uses the default stream when running callbacks, so we
+  // pass in the current CUDA stream in case it is not the default.
+  const auto& device = replicas_[0][0].device();
+  c10::DeviceType deviceType = device.type();
+  const c10::impl::VirtualGuardImpl guard =
+        c10::impl::VirtualGuardImpl{deviceType};
+  const c10::Stream currentStream =
+        guard.getStream(device);
+
+  // Run callback with the current stream
+  c10::OptionalStreamGuard currentStreamGuard{currentStream};
+
+  if (should_collect_runtime_stats()) {
+    record_backward_comm_start_time();
+  }
+
+  // launch all reduce local used map
+  all_reduce_local_used_map();
+
+  // prepare to set unused_parameters_, if it is static graph,
+  // unused_parameters_ will not change after 1st iteration.
+  if (static_graph_ && num_iterations_ == 1) {
+    unused_parameters_.clear();
+  }
+
+  // copy all gradients to buckets
+  for (size_t replica_index = 0; replica_index < replicas_.size();
+       replica_index++) {
+    for (size_t variable_index = 0; variable_index < replicas_[replica_index].size();
+         variable_index++) {
+      const auto index = VariableIndex(replica_index, variable_index);
+      // set unused_parameters_
+      if (static_graph_ && num_iterations_ == 1 && numGradHooksTriggeredMap_[index] == 0) {
```

Comment:
nit: It might be easier to index with `size_t variable_index` over `VariableIndex` structure since `replica_index` is always 0, will help the code be simpler